pkg/migrate/azure/privatedns.go

```diff
@@ -0,0 +1,744 @@
+	virtualNetworkLinksClient.Authorizer = session.Authorizer
+
+	virtualNetworksClient := aznetwork.NewVirtualNetworksClient(session.Credentials.SubscriptionID)
+	virtualNetworksClient.Authorizer = session.Authorizer
+
+	return &privateDNSClient{resourceGroup, vnetResourceGroup, virtualNetwork, zonesClient, recordsetsClient, virtualNetworkLinksClient, virtualNetworksClient}
+}
+
+// Gets a single private zone and serializes it
+func (client *privateDNSClient) GetZoneSerialized(privateZone string) (*privateDNSZoneSerialized, error) {
+	ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer cancel()
+
+	zone, err := client.zonesClient.Get(ctx, client.resourceGroup, privateZone)
+	zone.Response.Response = nil
+	if err != nil {
+		return nil, err
+	}
+
+	privateDNSZoneSerialized := privateDNSZoneSerialized{}
+	szone, err := serializeObject(zone)
+	if err != nil {
+		return nil, err
+	}
+	privateDNSZoneSerialized.zone = szone
+
+	for recordsetsPage, err := client.recordsetsClient.List(ctx, client.resourceGroup, to.String(zone.Name), to.Int32Ptr(100), ""); recordsetsPage.NotDone(); err = recordsetsPage.NextWithContext(ctx) {
+		if err != nil {
+			return nil, err
+		}
+
+		for _, rs := range recordsetsPage.Values() {
+			srs, err := serializeObject(rs)
+			if err != nil {
+				return nil, err
+			}
+			privateDNSZoneSerialized.recordsets = append(privateDNSZoneSerialized.recordsets, srs)
+		}
+	}
+
+	return &privateDNSZoneSerialized, nil
+}
+
+// Gets all private zones and serializes them
+func (client *privateDNSClient) GetZonesSerialized() ([]*privateDNSZoneSerialized, error) {
+	ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
+	defer cancel()
+
+	var privateDNSZonesSerialized []*privateDNSZoneSerialized
+	for zonesPage, err := client.zonesClient.List(ctx, to.Int32Ptr(100)); zonesPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
+		if err != nil {
+			return nil, err
+		}
+
+		for _, zone := range zonesPage.Values() {
+			zone.Response.Response = nil
+			szone, err := serializeObject(zone)
+			if err != nil {
+				return nil, err
+			}
+
+			privateDNSZoneSerialized := privateDNSZoneSerialized{}
+			privateDNSZoneSerialized.zone = szone
+
+			resourceGroup := idToResourceGroup(*zone.ID)
+			for recordsetsPage, err := client.recordsetsClient.List(ctx, resourceGroup, to.String(zone.Name), to.Int32Ptr(100), ""); recordsetsPage.NotDone(); err = recordsetsPage.NextWithContext(ctx) {
+				if err != nil {
+					return nil, err
+				}
+
+				for _, rs := range recordsetsPage.Values() {
+					srs, err := serializeObject(rs)
+					if err != nil {
+						return nil, err
+					}
+					privateDNSZoneSerialized.recordsets = append(privateDNSZoneSerialized.recordsets, srs)
+				}
+			}
+			privateDNSZonesSerialized = append(privateDNSZonesSerialized, &privateDNSZoneSerialized)
+		}
+	}
+
+	return privateDNSZonesSerialized, nil
+}
```

Comment:
1. this function is listing all the zones and then listing all recordsets for all these zones, when actually we shouldn't need to list records for zones we are not going to migrate.. so we should be able to reduce these API calls.

So i think this function should be reduce to listing the zones only. the returned value should provide
{ID, resourceGroupID, type}

2. I see no reason for all the serialize deserialize here, so we should drop them.

Reply:
As stated earlier, there was indeed an issue, and I can demonstrate it. While my curiosity wants to know what it is, it is no longer relevant to this code, so marking this as resolved.